orthanc/orthanc-anon/plugin/pixl.py

```diff
@@ -219,42 +216,151 @@ def OnHeartBeat(output, uri, **request) -> Any:  # noqa: ARG001
         study_identifiers = get_study_info(dataset)
         anonymise_and_validate_dicom(dataset, config_path=None, synchronise_pixl_db=True)
-        return orthanc.ReceivedInstanceAction.MODIFY, write_dataset_to_bytes(dataset)
+        return write_dataset_to_bytes(dataset)
     except PixlDiscardError as error:
-        logger.warning("Skipping {}: {}", study_identifiers, error)
-        return orthanc.ReceivedInstanceAction.DISCARD, None
-    except:  # noqa: E722 Allow bare except
-        # Called from a callback in orthanc so will never take down the service
-        # we want to do dicard anything on failure so we don't leak identifiable information
-        logger.exception("{} failed", study_identifiers)
-        return orthanc.ReceivedInstanceAction.DISCARD, None
+        logger.warning("Skipping instance for study {}: {}", study_identifiers, error)
+        raise
+
+
+def ImportStudyFromRaw(output, uri, **request):  # noqa: ARG001
+    """
+    Import a study from Orthanc Raw.
+
+    - Pull a study from Orthanc Raw based on its resource ID. Wait for the study to be stable.
+    - Iterate over instances and anonymise them
+    - Re-upload the study via the dicom-web api. Wait for the study to be stable.
+    - Notify the PIXL export-api to send the study the to relevant endpoint
+    """
+    payload = json.loads(request["body"])
+    study_uid = payload["StudyInstanceUID"]
+    query_id = payload["QueryID"]
+    retrieve_response = json.loads(
+        orthanc.RestApiPost(f"/queries/{query_id}/retrieve", json.dumps({}))
+    )
+    logger.debug(
+        "Response from retrieving study {} from Orthan Raw: {}", study_uid, retrieve_response
+    )
+
+    # Download the zipped study from Orthanc Anon
+    study_resource_id = _get_existing_study(study_uid=study_uid)
+    if study_resource_id is None:
+        return
+    zipped_study_bytes = BytesIO(orthanc.RestApiGet(f"/studies/{study_resource_id}/archive"))
+    logger.trace("Study data response {}", zipped_study_bytes)
+
+    # Delete the oringal study now in case anything goes wrong with the anonymisation.
+    # We don't want to leave the original (non-anonymised) study on Orthanc Anon
+    logger.info(
+        "Deleteing non-anonymised study with UID {} and resource ID {} from Orthanc Anon",
+        study_uid,
+        study_resource_id,
+    )
+    orthanc.RestApiDelete(f"/studies/{study_resource_id}")
+
+    # Anonymise the study, re-upload to Orthanc Anon, and notify the export API to export the study
+    with ZipFile(zipped_study_bytes) as zipped_study:
+        anonymised_instances_bytes, anonymised_study_uid = _anonymise_study_instances(
+            zipped_study=zipped_study,
+            study_uid=study_uid,
+        )
+    _upload_instances(anonymised_instances_bytes)
+    anonymised_study_resource_id = _get_existing_study(study_uid=anonymised_study_uid)
+    Send(study_id=anonymised_study_resource_id)
+
+
+def _get_existing_study(study_uid: str) -> Optional[str]:
```

Comment:
maybe this is clearer?
```suggestion
def _get_study_resource_id(study_uid: str) -> Optional[str]:
```

Reply:
yep, much clearer